### DISCUSSION TITLE
2d grid evaluation for collision checking

### DISCUSSION THREAD

**Discussion by kleindieck** [coin-or/CppAD#199] [Q&A]
Hi!
I am working on trajectory optimization and have a hard time to integrate collision checking given a 2d signed distance field, where each grid cell provides the distance to the closest obstacle.
Is there somewhere an example of 2d table interpolation? I only can find examples referencing the 1d case.

Best,
Jan

**Comment by bradbell** [comment]
Are you referring to the following example as the 1d case ?
https://cppad.readthedocs.io/latest/interp_onetape.cpp.html

**Reply by kleindieck** [reply]
Yes! A code snippet for a 2d grid would be great

**Comment by bradbell** [comment]
This would take me some time to do and I am currently working on other things. Perhaps you would like to to try with my help ?

**Reply by kleindieck** [reply]
Sure, I can put some work in this. Was just hoping that someone has used cppad for this use case before :)

**Comment by bradbell** [comment]
Take a look at
https://cppad.readthedocs.io/latest/atomic_four_bilinear.cpp.html

This uses atomic functions, instead of discrete functions, because the discrete functions only take one dimensional arguments.
I only implemented forward mode up to order 1. If you use a higher order forward, reverse mode,  sparsity, or base2ad, you will have to implement those.

**Reply by kleindieck** [reply]
Great, thanks. Will have a look

**Reply by kleindieck** [reply]
One more question: Is there an example somewhere on how do use the ADFun object when solving with IPOPT (In the FG_eval class)

**Comment by bradbell** [comment]
See the heading Example on
https://cppad.readthedocs.io/latest/ipopt_solve.html#example

**Comment by bradbell** [comment]
I suspect that you may need to define more of the virtual functions in atomic_bilinear to get it to work in this context.
